src/internal.c

```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
+ */
+static DtlsFragBucket* CombineFragBuckets(DtlsMsg* msg, DtlsFragBucket* cur,
+        DtlsFragBucket* next, word32 offset, const byte* data, word32 dataSz,
+        void* heap)
 {
-    WOLFSSL_ENTER("DtlsMsgSet()");
-    if (msg != NULL && data != NULL && msg->fragSz <= msg->sz &&
-        fragSz <= msg->sz && fragOffset <= msg->sz &&
-        (fragOffset + fragSz) <= msg->sz) {
-        DtlsFrag* cur = msg->fragList;
-        DtlsFrag* prev = cur;
-        DtlsFrag* newFrag;
-        word32 bytesLeft = fragSz; /* could be overlapping fragment */
-        word32 startOffset = fragOffset;
-        word32 added;
+    word32 offsetEnd = offset + dataSz;
+    word32 newOffset = min(cur->offset, offset);
+    word32 newOffsetEnd;
+    word32 newSz;
+    DtlsFragBucket** chosenBucket;
+    DtlsFragBucket* newBucket;
+    DtlsFragBucket* otherBucket;
+    byte combineNext = FALSE;
 
-        msg->seq = seq;
-        msg->epoch = epoch;
-        msg->type = type;
+    if (next != NULL && offsetEnd >= next->offset)
+        combineNext = TRUE;
 
-        if (fragOffset == 0) {
-            XMEMCPY(msg->buf, data - DTLS_HANDSHAKE_HEADER_SZ,
-                    DTLS_HANDSHAKE_HEADER_SZ);
-            c32to24(msg->sz, msg->msg - DTLS_HANDSHAKE_FRAG_SZ);
-        }
+    if (combineNext)
+        newOffsetEnd = next->offset + next->sz;
+    else
+        newOffsetEnd = max(cur->offset + cur->sz, offsetEnd);
 
-        /* if no message data, just return */
-        if (fragSz == 0)
-            return 0;
+    newSz = newOffsetEnd - newOffset;
 
-        /* if list is empty add full fragment to front */
-        if (cur == NULL) {
-            newFrag = CreateFragment(&fragOffset, fragOffset + fragSz - 1, data,
-                                     msg->msg, &bytesLeft, heap);
-            if (newFrag == NULL)
-                return MEMORY_E;
+    /* Expand the larger bucket if data bridges the gap between cur and next */
+    if (!combineNext || cur->sz >= next->sz) {
+        chosenBucket = &cur;
+        otherBucket = next;
+    }
+    else {
+        chosenBucket = &next;
+        otherBucket = cur;
+    }
+
+    {
+        DtlsFragBucket* tmp = (DtlsFragBucket*)XREALLOC(*chosenBucket,
+                sizeof(DtlsFragBucket) + newSz, heap, DYNAMIC_TYPE_DTLS_FRAG);
+        if (tmp == NULL)
+            return NULL;
+        newBucket = *chosenBucket = tmp;
+    }
 
-            msg->fragSz = fragSz;
-            msg->fragList = newFrag;
+    if (combineNext) {
+        /* Move data already in bucket first. Use memmove since newBucket may be
+         * next. */
+        XMEMMOVE(newBucket->buf + (next->offset - newOffset), next->buf,
+                next->sz);
+        /* memory after newOffsetEnd is already copied. Don't do extra work. */
+        newOffsetEnd = next->offset;
+    }
 
-            return 0;
+    if (newOffset == offset) {
+        /* data comes first */
+        if (newOffsetEnd <= offsetEnd) {
```

Comment:
nit: can this be < ?

Reply:
It can be if `data` extends into `next`. Copying from `next` modifies `newOffsetEnd`.